File: killrvideo-service-videocatalog/src/test/java/com/killrvideo/service/video/grpc/VideoCatalogServiceGrpcTest.java
package com.killrvideo.service.video.grpc;

import com.killrvideo.dse.dto.ResultListPage;
import com.killrvideo.dse.dto.Video;
import com.killrvideo.messaging.dao.MessagingDao;
import com.killrvideo.service.video.dto.LatestVideosPage;
import com.killrvideo.service.video.dto.UserVideo;
import com.killrvideo.service.video.request.GetLatestVideoPreviewsRequestData;
import com.killrvideo.service.video.request.GetUserVideoPreviewsRequestData;
import com.killrvideo.service.video.repository.VideoCatalogRepository;
import com.killrvideo.utils.GrpcMappingUtils;
import io.grpc.stub.StreamObserver;
import killrvideo.video_catalog.VideoCatalogServiceOuterClass.*;
import killrvideo.video_catalog.events.VideoCatalogEvents;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.Arrays;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

import static com.killrvideo.utils.GrpcMappingUtils.uuidToUuid;
import static java.util.Collections.singletonList;
import static org.mockito.Mockito.*;

@SuppressWarnings("unchecked")
class VideoCatalogServiceGrpcTest {
    @InjectMocks private VideoCatalogServiceGrpc service;
    @Mock
    private MessagingDao messagingDao;
    @Mock
    private VideoCatalogRepository videoCatalogRepository;
    @Mock
    private VideoCatalogServiceGrpcValidator validator;
    @Mock
    private VideoCatalogServiceGrpcMapper mapper;
    private AutoCloseable closeable;

    @BeforeEach
    public void openMocks() {
        closeable = MockitoAnnotations.openMocks(this);
    }

    @AfterEach
    public void releaseMocks() throws Exception {
        closeable.close();
    }

    @Test
    public void testSubmitYouTubeVideoWithValidationFailure() {
        SubmitYouTubeVideoRequest grpcReq = SubmitYouTubeVideoRequest.getDefaultInstance();
        StreamObserver<SubmitYouTubeVideoResponse > grpcResObserver = mock(StreamObserver.class);

        doThrow(new IllegalArgumentException()).when(this.validator)
                .validateGrpcRequest_submitYoutubeVideo(any(), any());

        Assertions.assertThrows(IllegalArgumentException.class, () ->
                service.submitYouTubeVideo(grpcReq, grpcResObserver)
        );
    }

    @Test
    public void testSubmitYouTubeVideoWithInsertFailure() {
        SubmitYouTubeVideoRequest grpcReq = SubmitYouTubeVideoRequest.getDefaultInstance();
        StreamObserver<SubmitYouTubeVideoResponse > grpcResObserver = mock(StreamObserver.class);

        doNothing().when(this.validator).validateGrpcRequest_submitYoutubeVideo(any(), any());
        Video video = mock(Video.class);
        when(this.mapper.mapSubmitYouTubeVideoRequestAsVideo(any())).thenReturn(video);

        when(this.videoCatalogRepository.insertVideoAsync(any())).thenReturn(
            CompletableFuture.failedFuture(new Exception())
        );

        service.submitYouTubeVideo(grpcReq, grpcResObserver);

        verify(grpcResObserver, times(1)).onError(any());
        verify(grpcResObserver, times(0)).onNext(any());
        verify(grpcResObserver, times(0)).onCompleted();
    }

    @Test
    public void testSubmitYouTubeVideoWithSendFailure() {
        SubmitYouTubeVideoRequest grpcReq = SubmitYouTubeVideoRequest.getDefaultInstance();
        StreamObserver<SubmitYouTubeVideoResponse > grpcResObserver = mock(StreamObserver.class);

        doNothing().when(this.validator).validateGrpcRequest_submitYoutubeVideo(any(), any());
        Video video = mock(Video.class);
        when(this.mapper.mapSubmitYouTubeVideoRequestAsVideo(any())).thenReturn(video);

        when(this.videoCatalogRepository.insertVideoAsync(any())).thenReturn(
                CompletableFuture.completedFuture(null)
        );
        VideoCatalogEvents.YouTubeVideoAdded event = VideoCatalogEvents.YouTubeVideoAdded.getDefaultInstance();
        when(this.mapper.createYouTubeVideoAddedEvent(any())).thenReturn(event);
        when(this.messagingDao.sendEvent(any(), any())).thenReturn(
                CompletableFuture.failedFuture(new Exception())
        );

        service.submitYouTubeVideo(grpcReq, grpcResObserver);

        verify(grpcResObserver, times(1)).onError(any());
        verify(grpcResObserver, times(0)).onNext(any());
        verify(grpcResObserver, times(0)).onCompleted();
    }

    @Test
    public void testSubmitYouTubeVideo() {
        SubmitYouTubeVideoRequest grpcReq = SubmitYouTubeVideoRequest.getDefaultInstance();
        StreamObserver<SubmitYouTubeVideoResponse > grpcResObserver = mock(StreamObserver.class);

        doNothing().when(this.validator).validateGrpcRequest_submitYoutubeVideo(any(), any());
        Video video = mock(Video.class);
        when(this.mapper.mapSubmitYouTubeVideoRequestAsVideo(any())).thenReturn(video);

        when(this.videoCatalogRepository.insertVideoAsync(any())).thenReturn(
                CompletableFuture.completedFuture(null)
        );
        VideoCatalogEvents.YouTubeVideoAdded event = VideoCatalogEvents.YouTubeVideoAdded.getDefaultInstance();
        when(this.mapper.createYouTubeVideoAddedEvent(any())).thenReturn(event);
        when(this.messagingDao.sendEvent(any(), any())).thenReturn(
                CompletableFuture.completedFuture(null)
        );

        service.submitYouTubeVideo(grpcReq, grpcResObserver);

        verify(grpcResObserver, times(0)).onError(any());
        verify(grpcResObserver, times(1)).onNext(any());
        verify(grpcResObserver, times(1)).onCompleted();
    }

    @Test
    void testGetLatestVideoPreviewsWithValidationFailure() {
        GetLatestVideoPreviewsRequest grpcReq = GetLatestVideoPreviewsRequest.getDefaultInstance();
        StreamObserver<GetLatestVideoPreviewsResponse> grpcResObserver = mock(StreamObserver.class);

        doThrow(new IllegalArgumentException()).when(this.validator)
                .validateGrpcRequest_getLatestPreviews(any(), any());

        Assertions.assertThrows(IllegalArgumentException.class, () ->
                service.getLatestVideoPreviews(grpcReq, grpcResObserver)
        );
    }

    @Test
    void testGetLatestVideoPreviewsWithQueryFailure() {
        GetLatestVideoPreviewsRequest grpcReq = GetLatestVideoPreviewsRequest.getDefaultInstance();
        StreamObserver<GetLatestVideoPreviewsResponse> grpcResObserver = mock(StreamObserver.class);

        doNothing().when(this.validator).validateGrpcRequest_getLatestPreviews(any(), any());

        GetLatestVideoPreviewsRequestData requestData = mock(GetLatestVideoPreviewsRequestData.class);
        when(this.mapper.parseGetLatestVideoPreviewsRequest(any(), any())).thenReturn(requestData);

        when(videoCatalogRepository.getLatestVideoPreviewsAsync(any())).thenReturn(
                CompletableFuture.failedFuture(new RuntimeException())
        );

        service.getLatestVideoPreviews(grpcReq, grpcResObserver);

        verify(grpcResObserver, times(1)).onError(any());
        verify(grpcResObserver, times(0)).onNext(any());
        verify(grpcResObserver, times(0)).onCompleted();
    }

    @Test
    void testGetLatestVideoPreviews() {
        GetLatestVideoPreviewsRequest grpcReq = GetLatestVideoPreviewsRequest.getDefaultInstance();
        StreamObserver<GetLatestVideoPreviewsResponse> grpcResObserver = mock(StreamObserver.class);

        doNothing().when(this.validator).validateGrpcRequest_getLatestPreviews(any(), any());

        GetLatestVideoPreviewsRequestData requestData = mock(GetLatestVideoPreviewsRequestData.class);
        when(this.mapper.parseGetLatestVideoPreviewsRequest(any(), any())).thenReturn(requestData);

        LatestVideosPage returnedPage = mock(LatestVideosPage.class);
        when(videoCatalogRepository.getLatestVideoPreviewsAsync(any())).thenReturn(
                CompletableFuture.completedFuture(returnedPage)
        );

        GetLatestVideoPreviewsResponse response = GetLatestVideoPreviewsResponse.getDefaultInstance();
        when(this.mapper.mapLatestVideoToGrpcResponse(any())).thenReturn(response);

        service.getLatestVideoPreviews(grpcReq, grpcResObserver);

        verify(grpcResObserver, times(0)).onError(any());
        verify(grpcResObserver, times(1)).onNext(any());
        verify(grpcResObserver, times(1)).onCompleted();
    }

    @Test
    void testGetVideoWithValidationFailure() {
        GetVideoRequest grpcReq = GetVideoRequest.getDefaultInstance();
        StreamObserver<GetVideoResponse> grpcResObserver = mock(StreamObserver.class);

        doThrow(new IllegalArgumentException()).when(this.validator)
                .validateGrpcRequest_getVideo(any(), any());

        Assertions.assertThrows(IllegalArgumentException.class, () ->
                service.getVideo(grpcReq, grpcResObserver)
        );
    }

    @Test
    void testGetVideoWithQueryFailure() {
        GetVideoRequest grpcReq = getVideoRequest(UUID.randomUUID());
        StreamObserver<GetVideoResponse> grpcResObserver = mock(StreamObserver.class);

        doNothing().when(this.validator).validateGrpcRequest_getVideo(any(), any());
        when(videoCatalogRepository.getVideoById(any())).thenReturn(
                CompletableFuture.failedFuture(new Exception())
        );

        service.getVideo(grpcReq, grpcResObserver);
        verify(grpcResObserver, times(1)).onError(any());
        verify(grpcResObserver, times(0)).onNext(any());
        verify(grpcResObserver, times(0)).onCompleted();
    }

    @Test
    void testGetVideoWithNoVideoFound() {
        GetVideoRequest grpcReq = getVideoRequest(UUID.randomUUID());
        StreamObserver<GetVideoResponse> grpcResObserver = mock(StreamObserver.class);

        doNothing().when(this.validator).validateGrpcRequest_getVideo(any(), any());
        when(videoCatalogRepository.getVideoById(any())).thenReturn(
                CompletableFuture.completedFuture(null)
        );

        service.getVideo(grpcReq, grpcResObserver);
        verify(grpcResObserver, times(1)).onError(any());
        verify(grpcResObserver, times(0)).onNext(any());
        verify(grpcResObserver, times(0)).onCompleted();
    }

    @Test
    void testGetVideo() {
        GetVideoRequest grpcReq = getVideoRequest(UUID.randomUUID());
        StreamObserver<GetVideoResponse> grpcResObserver = mock(StreamObserver.class);

        doNothing().when(this.validator).validateGrpcRequest_getVideo(any(), any());
        Video video = mock(Video.class);
        when(videoCatalogRepository.getVideoById(any())).thenReturn(
                CompletableFuture.completedFuture(video)
        );
        GetVideoResponse response = GetVideoResponse.getDefaultInstance();
        when(this.mapper.mapFromVideotoVideoResponse(any())).thenReturn(response);

        service.getVideo(grpcReq, grpcResObserver);
        verify(grpcResObserver, times(0)).onError(any());
        verify(grpcResObserver, times(1)).onNext(any());
        verify(grpcResObserver, times(1)).onCompleted();
    }

    @Test
    void testGetVideoPreviewsWithValidationFailure() {
        GetVideoPreviewsRequest grpcReq = GetVideoPreviewsRequest.getDefaultInstance();
        StreamObserver<GetVideoPreviewsResponse> grpcResObserver = mock(StreamObserver.class);

        doThrow(new IllegalArgumentException()).when(this.validator)
                .validateGrpcRequest_getVideoPreviews(any(), any());

        Assertions.assertThrows(IllegalArgumentException.class, () ->
                service.getVideoPreviews(grpcReq, grpcResObserver)
        );
    }

    @Test
    void testGetVideoPreviewsWithoutVideioId() {
        GetVideoPreviewsRequest grpcReq = getVideoPreviewsRequest();
        StreamObserver<GetVideoPreviewsResponse> grpcResObserver = mock(StreamObserver.class);

        doNothing().when(this.validator).validateGrpcRequest_getVideoPreviews(any(), any());
        service.getVideoPreviews(grpcReq, grpcResObserver);

        verify(grpcResObserver, times(0)).onError(any());
        verify(grpcResObserver, times(1)).onNext(any());
        verify(grpcResObserver, times(1)).onCompleted();
    }

    @Test
    void testGetVideoPreviewsWithQueryFailure() {
        UUID videoid = UUID.randomUUID();
        GetVideoPreviewsRequest grpcReq = getVideoPreviewsRequest(videoid);
        StreamObserver<GetVideoPreviewsResponse> grpcResObserver = mock(StreamObserver.class);

        doNothing().when(this.validator).validateGrpcRequest_getVideoPreviews(any(), any());
        when(this.videoCatalogRepository.getVideoPreview(any())).thenReturn(
            CompletableFuture.failedFuture(new Exception())
        );
        service.getVideoPreviews(grpcReq, grpcResObserver);

        verify(grpcResObserver, times(1)).onError(any());
        verify(grpcResObserver, times(0)).onNext(any());
        verify(grpcResObserver, times(0)).onCompleted();
    }

    @Test
    void testGetVideoPreviews() {
        UUID videoid = UUID.randomUUID();
        GetVideoPreviewsRequest grpcReq = getVideoPreviewsRequest(videoid);
        StreamObserver<GetVideoPreviewsResponse> grpcResObserver = mock(StreamObserver.class);

        doNothing().when(this.validator).validateGrpcRequest_getVideoPreviews(any(), any());
        Video video = mock(Video.class);
        List<Video> videos = singletonList(video);
        when(this.videoCatalogRepository.getVideoPreview(any())).thenReturn(
                CompletableFuture.completedFuture(videos)
        );

        GetVideoPreviewsResponse response = GetVideoPreviewsResponse.getDefaultInstance();
        when(this.mapper.mapToGetVideoPreviewsResponse(any())).thenReturn(response);

        service.getVideoPreviews(grpcReq, grpcResObserver);

        verify(grpcResObserver, times(0)).onError(any());
        verify(grpcResObserver, times(1)).onNext(any());
        verify(grpcResObserver, times(1)).onCompleted();
    }

    @Test
    void testGetUserVideoPreviewsWithValidationFailure() {
        GetUserVideoPreviewsRequest grpcReq = GetUserVideoPreviewsRequest.getDefaultInstance();
        StreamObserver<GetUserVideoPreviewsResponse> grpcResObserver = mock(StreamObserver.class);

        doThrow(new IllegalArgumentException()).when(this.validator)
                .validateGrpcRequest_getUserVideoPreviews(any(), any());

        Assertions.assertThrows(IllegalArgumentException.class, () ->
                service.getUserVideoPreviews(grpcReq, grpcResObserver)
        );
    }

    @Test
    void testGetUserVideoPreviewsWithQueryFailure() {
        UUID userid = UUID.randomUUID();
        GetUserVideoPreviewsRequest grpcReq = getUserVideoPreviewsRequest(userid);
        StreamObserver<GetUserVideoPreviewsResponse> grpcResObserver = mock(StreamObserver.class);

        doNothing().when(this.validator).validateGrpcRequest_getUserVideoPreviews(any(), any());

        GetUserVideoPreviewsRequestData requestData = getUserVideoPreviewsRequestData(userid);
        when(this.mapper.parseGetUserVideoPreviewsRequest(grpcReq)).thenReturn(requestData);

        when(this.videoCatalogRepository.getUserVideosPreview(any())).thenReturn(
                CompletableFuture.failedFuture(new Exception())
        );

        service.getUserVideoPreviews(grpcReq, grpcResObserver);
        verify(grpcResObserver, times(1)).onError(any());
        verify(grpcResObserver, times(0)).onNext(any());
        verify(grpcResObserver, times(0)).onCompleted();
    }

    @Test
    void testGetUserVideoPreviews() {
        UUID userid = UUID.randomUUID();
        GetUserVideoPreviewsRequest grpcReq = getUserVideoPreviewsRequest(userid);
        StreamObserver<GetUserVideoPreviewsResponse> grpcResObserver = mock(StreamObserver.class);

        doNothing().when(this.validator).validateGrpcRequest_getUserVideoPreviews(any(), any());

        GetUserVideoPreviewsRequestData requestData = getUserVideoPreviewsRequestData(userid);
        when(this.mapper.parseGetUserVideoPreviewsRequest(grpcReq)).thenReturn(requestData);

        ResultListPage<UserVideo> resultListPage = mock(ResultListPage.class);

        when(this.videoCatalogRepository.getUserVideosPreview(any())).thenReturn(
                CompletableFuture.completedFuture(resultListPage)
        );
        GetUserVideoPreviewsResponse response = GetUserVideoPreviewsResponse.getDefaultInstance();
        when(this.mapper.mapToGetUserVideoPreviewsResponse(any(), any())).thenReturn(response);

        service.getUserVideoPreviews(grpcReq, grpcResObserver);
        verify(grpcResObserver, times(0)).onError(any());
        verify(grpcResObserver, times(1)).onNext(any());
        verify(grpcResObserver, times(1)).onCompleted();
    }

    private GetVideoRequest getVideoRequest(UUID videoid) {
        return GetVideoRequest.newBuilder()
                .setVideoId(uuidToUuid(videoid))
                .build();
    }

    private GetVideoPreviewsRequest getVideoPreviewsRequest(UUID... videoIds) {
        return GetVideoPreviewsRequest.newBuilder()
                .addAllVideoIds(
                        Arrays.stream(videoIds).map(GrpcMappingUtils::uuidToUuid).collect(Collectors.toList())
                )
                .build();
    }

    private GetUserVideoPreviewsRequest getUserVideoPreviewsRequest(UUID userid) {
        return GetUserVideoPreviewsRequest.newBuilder()
                .setUserId(uuidToUuid(userid))
                .build();
    }

    private GetUserVideoPreviewsRequestData getUserVideoPreviewsRequestData(UUID userid) {
        return new GetUserVideoPreviewsRequestData(userid);
    }
}

File: killrvideo-service-videocatalog/src/test/kotlin/com/killrvideo/service/video/grpc/VideoCatalogServiceGrpcTest.kt
package com.killrvideo.service.video.grpc

import com.killrvideo.dse.dto.ResultListPage
import com.killrvideo.exception.NotFoundException
import com.killrvideo.messaging.dao.MessagingDao
import com.killrvideo.service.video.dto.LatestVideosPage
import com.killrvideo.service.video.dto.UserVideo
import com.killrvideo.service.video.dto.Video
import com.killrvideo.service.video.repository.CachedVideoCatalogRepository
import com.killrvideo.service.video.request.GetUserVideoPreviewsRequestData
import com.killrvideo.utils.GrpcMappingUtils.randomUuid
import com.killrvideo.utils.GrpcMappingUtils.uuidToUuid
import io.mockk.*
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import killrvideo.video_catalog.*
import killrvideo.video_catalog.VideoCatalogServiceOuterClass.*
import killrvideo.video_catalog.events.VideoCatalogEvents.YouTubeVideoAdded
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.util.*
import java.util.concurrent.CompletableFuture

internal class VideoCatalogServiceGrpcTest {
    @InjectMockKs
    private lateinit var service: VideoCatalogServiceGrpc

    @MockK
    private lateinit var messagingDao: MessagingDao

    @MockK
    private lateinit var videoCatalogRepository: CachedVideoCatalogRepository

    @MockK
    private lateinit var validator: VideoCatalogServiceGrpcValidator

    @MockK
    private lateinit var mapper: VideoCatalogServiceGrpcMapper

    val serviceKey ="VideoCatalogService"
    private val topicVideoCreated = "topic-kv-videoCreation"

    @BeforeEach
    fun setUp() = MockKAnnotations.init(this, relaxUnitFun = true)

    @Test
    fun testSubmitYouTubeVideoWithValidationFailure() {
        val request = createSubmitYouTubeVideoRequest()
        every { validator.validateGrpcRequest_submitYoutubeVideo(any()) } throws IllegalArgumentException()
        assertThrows<IllegalArgumentException> {
            runBlocking { service.submitYouTubeVideo(request) }
        }
    }

    @Test
    fun testSubmitYouTubeVideoWithInsertFailure() {
        val request = createSubmitYouTubeVideoRequest()
        every { validator.validateGrpcRequest_submitYoutubeVideo(any()) } just Runs

        coEvery { videoCatalogRepository.insertVideoAsync(any()) } throws Exception()

        assertThrows<Exception> {
            runBlocking { service.submitYouTubeVideo(request) }
        }
    }

    private fun createSubmitYouTubeVideoRequest() = submitYouTubeVideoRequest {
        userId = randomUuid()
        videoId = randomUuid()
    }

    @Test
    fun testSubmitYouTubeVideoWithSendFailure() {
        val request = createSubmitYouTubeVideoRequest()
        every { validator.validateGrpcRequest_submitYoutubeVideo(any()) } just Runs
        coJustRun { videoCatalogRepository.insertVideoAsync(any()) }
        val event = YouTubeVideoAdded.getDefaultInstance()
        every { mapper.createYouTubeVideoAddedEvent(any()) } returns event
        val future: CompletableFuture<Void> = CompletableFuture.failedFuture(Exception())
        every { messagingDao.sendEvent(any(), any()) } returns future

        assertThrows<Exception> {
            runBlocking { service.submitYouTubeVideo(request) }
        }
    }

    @Test
    fun testSubmitYouTubeVideo() {
        val request = createSubmitYouTubeVideoRequest()
        every { validator.validateGrpcRequest_submitYoutubeVideo(any()) } just Runs
        coJustRun { videoCatalogRepository.insertVideoAsync(any()) }
        val event = YouTubeVideoAdded.getDefaultInstance()
        every { mapper.createYouTubeVideoAddedEvent(any()) } returns event
        every { messagingDao.sendEvent(any(), any()) } returns
                CompletableFuture.completedFuture(null)

        runBlocking { service.submitYouTubeVideo(request) }
        coVerify {
            videoCatalogRepository.insertVideoAsync(any())
        }

        verify {
            messagingDao.sendEvent(any(), any())
        }
    }

    @Test
    fun testGetLatestVideoPreviewsWithValidationFailure() {
        val request = getLatestVideoPreviewsRequest {}
        every { validator.validateGrpcRequest_getLatestPreviews(any()) } throws IllegalArgumentException()
        assertThrows<IllegalArgumentException> {
            runBlocking { service.getLatestVideoPreviews(request) }
        }
    }

    @Test
    fun testGetLatestVideoPreviewsWithQueryFailure() {
        val request = getLatestVideoPreviewsRequest {}
        every { validator.validateGrpcRequest_getLatestPreviews(any()) } just Runs
        coEvery { videoCatalogRepository.getLatestVideoPreviewsAsync(any()) } throws Exception()
        assertThrows<Exception> {
            runBlocking { service.getLatestVideoPreviews(request) }
        }
    }

    @Test
    fun testGetLatestVideoPreviews() {
        val request = getLatestVideoPreviewsRequest {}
        every { validator.validateGrpcRequest_getLatestPreviews(any()) } just Runs
        val returnedPage = mockk<LatestVideosPage>()
        coEvery { videoCatalogRepository.getLatestVideoPreviewsAsync(any()) } returns returnedPage
        val response = getLatestVideoPreviewsResponse {}
        every { mapper.mapLatestVideoToGrpcResponse(any()) } returns response
        runBlocking { service.getLatestVideoPreviews(request) }
    }

    @Test
    fun testGetVideoWithValidationFailure() {
        val request = getVideoRequest {}
        every { validator.validateGrpcRequest_getVideo(any()) } throws IllegalArgumentException()
        assertThrows<IllegalArgumentException> {
            runBlocking { service.getVideo(request) }
        }
    }

    @Test
    fun testGetVideoWithQueryFailure() {
        val request = createGetVideoRequest(UUID.randomUUID())
        every { validator.validateGrpcRequest_getVideo(any()) } just Runs
        coEvery { videoCatalogRepository.getVideoById(any()) } throws Exception()

        assertThrows<Exception> {
            runBlocking { service.getVideo(request) }
        }
    }

    @Test
    fun testGetVideoWithNoVideoFound() {
        val request = createGetVideoRequest(UUID.randomUUID())
        every { validator.validateGrpcRequest_getVideo(any()) } just Runs
        coEvery { videoCatalogRepository.getVideoById(any()) } returns null
        assertThrows<NotFoundException> {
            runBlocking { service.getVideo(request) }
        }
    }

    @Test
    fun testGetVideo() {
        val request = createGetVideoRequest(UUID.randomUUID())
        every { validator.validateGrpcRequest_getVideo(any()) } just Runs
        val video = mockk<Video>()
        coEvery { videoCatalogRepository.getVideoById(any()) } returns video

        val response = GetVideoResponse.getDefaultInstance()
        every { mapper.mapFromVideotoVideoResponse(any()) } returns response
        val result = runBlocking { service.getVideo(request) }
        assertEquals(response, result)
    }

    @Test
    fun testGetVideoPreviewsWithValidationFailure() {
        val request = GetVideoPreviewsRequest.getDefaultInstance()
        every { validator.validateGrpcRequest_getVideoPreviews(any()) } throws IllegalArgumentException()
        assertThrows<IllegalArgumentException> {
            runBlocking { service.getVideoPreviews(request) }
        }
    }

    @Test
    fun testGetVideoPreviewsWithoutVideioId() {
        val request = createGetVideoPreviewsRequest()
        every { validator.validateGrpcRequest_getVideoPreviews(any()) } just Runs
        val result = runBlocking { service.getVideoPreviews(request) }
        assertEquals(0, result.videoPreviewsCount)
    }

    @Test
    fun testGetVideoPreviewsWithQueryFailure() {
        val videoid = UUID.randomUUID()
        val request = createGetVideoPreviewsRequest(videoid)
        every { validator.validateGrpcRequest_getVideoPreviews(any()) } just Runs
        coEvery { videoCatalogRepository.getVideoPreview(any()) } throws Exception()

        assertThrows<Exception> {
            runBlocking { service.getVideoPreviews(request) }
        }
    }

    @Test
    fun testGetVideoPreviews() {
        val videoid = UUID.randomUUID()
        val request = createGetVideoPreviewsRequest(videoid)
        every { validator.validateGrpcRequest_getVideoPreviews(any()) } just Runs
        val video = mockk<Video>()
        val videos = listOf(video)
        coEvery { videoCatalogRepository.getVideoPreview(any()) } returns videos
        val response = GetVideoPreviewsResponse.getDefaultInstance()
        every { mapper.mapToGetVideoPreviewsResponse(any()) } returns response
        val result = runBlocking { service.getVideoPreviews(request) }
        assertEquals(response, result)
    }

    @Test
    fun testGetUserVideoPreviewsWithValidationFailure() {
        val request = GetUserVideoPreviewsRequest.getDefaultInstance()
        every { validator.validateGrpcRequest_getUserVideoPreviews(any()) } throws IllegalArgumentException()
        assertThrows<IllegalArgumentException> {
            runBlocking { service.getUserVideoPreviews(request) }
        }
    }

    @Test
    fun testGetUserVideoPreviewsWithQueryFailure() {
        val userid = UUID.randomUUID()
        val request = createGetUserVideoPreviewsRequest(userid)
        every { validator.validateGrpcRequest_getUserVideoPreviews(any()) } just Runs
        coEvery { videoCatalogRepository.getUserVideosPreview(any()) } throws Exception()
        assertThrows<Exception> {
            runBlocking { service.getUserVideoPreviews(request) }
        }
    }

    @Test
    fun testGetUserVideoPreviews() {
        val userid = UUID.randomUUID()
        val request = createGetUserVideoPreviewsRequest(userid)
        every { validator.validateGrpcRequest_getUserVideoPreviews(any()) } just Runs
        val resultListPage: ResultListPage<UserVideo> = mockk()
        coEvery { videoCatalogRepository.getUserVideosPreview(any()) } returns resultListPage
        val response = GetUserVideoPreviewsResponse.getDefaultInstance()
        every {
            mapper.mapToGetUserVideoPreviewsResponse(any(), any())
        } returns response
        val result = runBlocking { service.getUserVideoPreviews(request) }
        assertEquals(response, result)
    }

    private fun createGetVideoRequest(videoid: UUID): GetVideoRequest =
        getVideoRequest {
            videoId = uuidToUuid(videoid)
        }

    private fun createGetVideoPreviewsRequest(vararg videoids: UUID): GetVideoPreviewsRequest =
        getVideoPreviewsRequest {
            Arrays.stream(videoids).map { uuid: UUID -> uuidToUuid(uuid) }
                .forEach { videoIds.add(it) }
        }

    private fun createGetUserVideoPreviewsRequest(userid: UUID): GetUserVideoPreviewsRequest =
        getUserVideoPreviewsRequest {
            userId = uuidToUuid(userid)
        }

    private fun createGetUserVideoPreviewsRequestData(userid: UUID): GetUserVideoPreviewsRequestData =
        GetUserVideoPreviewsRequestData(userid)

}
